### DISCUSSION TITLE
Options as record type in version 6 is bad idea

### DISCUSSION THREAD

**Discussion by mssucksalot** [linq2db/linq2db#4704] [General]
If the options type is a record, the following syntax is not possible:

```C#
BulkCopyOptions options = new BulkCopyOptions();
options.MaxBatchSize = batch;
options.BulkCopyTimeout = 600; // 10 minutes
options.BulkCopyType = BulkCopyType.ProviderSpecific;
options.CheckConstraints = true;
options.FireTriggers = true;
```

This approach makes sense for chaining settings and (or) may be preferred by some developers.

If you are targeting .NET Standard 2.0 or .NET Framework, you will encounter the exception: `System.MissingMethodException: Method not found: 'Void LinqToDB.Data.BulkCopyOptions.set_MaxBatchSize(System.Nullable1<Int32>)'` with the following initialization:

```C#
BulkCopyOptions options = new BulkCopyOptions
{
    MaxBatchSize = batch,
    BulkCopyTimeout = 600, // 10 minutes
    BulkCopyType = BulkCopyType.ProviderSpecific,
    CheckConstraints = true,
    FireTriggers = true
};
```

The only way to initialize the options correctly is as follows:

```C#
BulkCopyOptions options = new BulkCopyOptions(
    MaxBatchSize: batch,
    BulkCopyTimeout: 600, // 10 minutes
    BulkCopyType: BulkCopyType.ProviderSpecific,
    CheckConstraints: true,
    FireTriggers: true
);
```
This is more of a workaround and highlights what appears to be a significant flaw. Each new version of C# introduces more questionable language "innovations," which requires caution in their use.

**Comment by igor-tkachev** [comment]
We have to make sure that an Option object is immutable. There is a lot of optimizations based on this.

**Comment by MaceWindu** [comment]
I would say we will not revert this change as:
1. This change was introduced in v5, not v6, so it is here already for some time
2. It is possible to use never C# versions in .net framework by setting language version explicitly

**Comment by mssucksalot** [comment]
@MaceWindu 

> This change was introduced in v5, not v6, so it is here already for some time

That’s actually the reason I maintained my own fork for the 2.x-3.x versions, to avoid frequent changes and releases. Sometimes this leads to issues, regressions, or problems like this one.

I had originally planned to upgrade to version 5, but since version 6.0 is almost ready, I decided to give it a try.

> It is possible to use never C# versions in .net framework by setting language version explicitly

I have tried that before issue is posted, but it doesn’t work - exactly as described above. Setting the language version is not enough.

**Comment by mssucksalot** [comment]
I see what you talked about. I need to set lang spec for my project referencing linq2db. Finally works, not ideally but at least covers 2nd and 3rd samples.

**Comment by mssucksalot** [comment]
@igor-tkachev  ---> System.MissingMethodException: Method not found: 'Void LinqToDB.Data.BulkCopyOptions.set_MaxBatchSize(System.Nullable`1<Int32>)'.

There were much more better ways to get it immutable.
